privacyidea/api/lib/utils.py

```diff
@@ -247,19 +247,44 @@ def verify_auth_token(auth_token, required_role=None):
     :param required_role: list of "user" and "admin"
     :return: dict with authtype, realm, rights, role, username, exp, nonce
     """
+    r = None
     if required_role is None:
         required_role = ["admin", "user"]
     if auth_token is None:
         raise AuthError(_("Authentication failure. Missing Authorization header."),
                         id=ERROR.AUTHENTICATE_AUTH_HEADER)
-    try:
-        r = jwt.decode(auth_token, current_app.secret_key, algorithms=['HS256'])
-    except jwt.DecodeError as err:
-        raise AuthError(_("Authentication failure. Error during decoding your token: {0!s}").format(err),
-                        id=ERROR.AUTHENTICATE_DECODING_ERROR)
-    except jwt.ExpiredSignature as err:
-        raise AuthError(_("Authentication failure. Your token has expired: {0!s}").format(err),
-                        id=ERROR.AUTHENTICATE_TOKEN_EXPIRED)
+
+    headers = jwt.get_unverified_header(auth_token)
+    if headers.get("alg") == "RS256":
+        # The trusted JWTs are RSA signed
+        trusted_jwts = current_app.config.get("PI_TRUSTED_JWT", [])
+        for trusted_jwt in trusted_jwts:
+            try:
+                if trusted_jwt.get("algorithm") == "RS256":
+                    j = jwt.decode(auth_token,
+                                   trusted_jwt.get("public_key"),
+                                   algorithms=["RS256"])
+                    if j == dict((k, trusted_jwt.get(k)) for k in ("role", "user", "resolver", "realm")):
```

Comment:
I think we'll run into the unordered dictionary issue here.
Also this would mean, that the payload must not contain more elements than specified in ``PI_TRUSTED_JWT``.
We could use python sets to solve both: ``if {"role", "user", "resolver", "realm"} <= six.viewkeys(j)``

Reply:
Hm, but in python2 and python3 two dicts compare to true:

    {"H": "b", 1:3} == {1:3, "H": "b"}

And this is what we want here.

Reply:
done in 41a29da1
This works for a lot of dicts:

~~~~python
>>> d1 = {"hallo": "du", 1:3, 17:2}
>>> d2 = {1:3, "hallo": "du"}
>>> d1 == d2
False
>>> dict((k, d1.get(k)) for k in ("hallo", 1)) == dict((k, d2.get(k)) for k in ("hallo", 1))
True
~~~~